src/utils/XFDR.cpp

```diff
@@ -0,0 +1,1095 @@
+
+    // Number of ranks used
+    static const UInt n_rank;
+
+    // Constants related to particular xl classes
+    static const String xlclass_intradecoys; // intradecoys
+    static const String xlclass_fulldecoysintralinks; // fulldecoysintralinks
+    static const String xlclass_interdecoys; // interdecoys
+    static const String xlclass_fulldecoysinterlinks; // fulldecoysinterlinks
+    static const String xlclass_intralinks; // intralinks
+    static const String xlclass_interlinks; // interlinks
+    static const String xlclass_monolinks;  // monolinks
+    static const String xlclass_monodecoys; // monodecoys
+    static const String xlclass_decoys; // decoys
+    static const String xlclass_hybriddecoysintralinks; // hybriddecoysintralinks
+    static const String xlclass_hybriddecoysinterlinks; // hybriddecoysintralinks
+
+    // Score range for calculating the FPs
+    static const double fpnum_score_step;
+
+    TOPPXFDR() :
+      TOPPBase("XFDR", "Calculates false discovery rate estimates on cross-link identifications", false)
+    {
+    }
+
+  protected:
+
+    /**
+     * @brief Used to define how PeptideIdentification are to be sorted based on some meta value (usually some score)
+     */
+    struct less_than_by_key
+    {
+      //Indizes of the rank one elements within the all_ids vector
+      const vector < PeptideIdentification > & all_ids;
+      const vector< UInt >  & rank_one_ids;
+      const String & meta_value;  // Meta value to sort the peptideIdentifications
+
+      inline bool operator()(const UInt & index1, const UInt & index2)
+      {
+        return (  TOPPXFDR::getXLMetaValue<double>(meta_value, all_ids[rank_one_ids[index1]], true)
+                > TOPPXFDR::getXLMetaValue<double>(meta_value, all_ids[rank_one_ids[index2]], true));
+      }
+    };
+
+    /**
+     * @brief Tests for descending ordering of a vector of rank one ids based on the score. Only used for assertions.
+     * @param order vector of ranks
+     * @param all_ids Vector with the actual PeptideIdentifictions which are queried for the score.
+     * @param rank_one_ids Vector of indices pointing to the rank one IDs in all_ids.
+     * @return Whether rank vector order reflects descending ordering of rank_one_ids by score.
+     */
+    static bool isSortedDescending(vector< UInt > & order, vector< PeptideIdentification > & all_ids, vector< UInt > & rank_one_ids )
+    {
+      for (Size i = 0; i < order.size() - 1; ++i)
+      {
+        if (  getXLMetaValue<double>("OpenXQuest:score", all_ids[rank_one_ids[order[i]]], true)
+            < getXLMetaValue<double>("OpenXQuest:score", all_ids[rank_one_ids[order[i + 1]]], true))
+        {
+          return false;
+        }
+      }
+      return true;
+    }
+
+    /**
+     * Searches for a meta value in the cross-link peptide ID. First in PeptideIdentification itself, then
+     * in the associated Peptide hits.
+     */
+    template<typename T>
```

Comment:
this method looks a bit fishy ;) is the template required? the logic is also a bit inexpected (e.g., if size == 2)
Consider reworking it (e.g., split it into 3 members)